src/range.cr

```diff
@@ -626,6 +626,101 @@ struct Range(B, E)
       end
     end
   end
+
+  # Clamps `self` within the boundaries of `range`.
+  #
+  # The result is the intersection of both ranges.
+  #
+  # ```
+  # (-5..5).clamp(0..100) # => 0..5
+  # (-5..5).clamp(-10..2) # => -5..2
+  # (-5..5).clamp(0..nil) # => 0..5
+  # ```
+  #
+  # If either range is exclusive, the result is exclusive, too.
+  # If both ranges are disparate, or have different `#direction` the result is
+  # an empty range expressed as `begin...begin` (or `end...end` if `begin` is
+  # `nil`).
+  #
+  # ```
+  # (-5...5).clamp(0..100) # => 0...5
+  # (-5..5).clamp(0...100) # => 0...5
+  # (-5..5).clamp(10..20)  # => -5...-5
+  # (-5..5).clamp(5..-5)   # => -5...-5
+  # (nil..0).clamp(1..nil) # => 0...0
+  # ```
+  def clamp(range : Range) : Range
+    range_direction = range.direction
+    direction = self.direction
+
+    if direction != range_direction && !direction.zero? && !range_direction.zero?
+      # result is an empty range and gets normalized
+      v = (self.begin || self.end).not_nil!
+      return v...v
+    end
+
+    if range_direction < 0
+      # if both ranges are descending, reverse range in order to work for `Number#clamp` calls
+      range = range.reverse
```

Comment:
I think we should always stay under the assumption that the range is ascending. If end is before the begin then the range is empty and it doesn't matter too much what happens to it then.